### ISSUE TITLE
alibi的mask和论文不一致

### ISSUE THREAD

**Issue by ReactiveCJ** [baichuan-inc/Baichuan-13B#172]
`def _gen_alibi_mask(n_head, max_pos):
    """used in inference only"""
    slopes = torch.Tensor(_get_interleave(n_head))
    alibi = slopes.unsqueeze(1).unsqueeze(1) * torch.arange(max_pos).unsqueeze(0).unsqueeze(0).expand(
        n_head, -1, -1)
    alibi = alibi.view(n_head, 1, max_pos)
    alibi_mask = torch.triu(
        _fill_with_neg_inf(torch.zeros([max_pos, max_pos])), 1
    )
    alibi_mask = alibi_mask.unsqueeze(0) + alibi
    return alibi_mask

def _buffered_future_mask(tensor, maxpos, alibi, attn_heads):
    """used in training only"""
    dim = tensor.size(1)
    _future_mask = torch.triu(
        _fill_with_neg_inf(torch.zeros([maxpos, maxpos])), 1
    )   
    _future_mask = _future_mask.unsqueeze(0) + alibi
    _future_mask = _future_mask.to(tensor)
    return _future_mask[:tensor.shape[0] * attn_heads, :maxpos, :maxpos]`

1. alibi的mask和论文不一致
2. 为何很多乘法计算之前，都要加上expand，感觉没有必要

**Comment by shiqingzhangCSU** [comment]
不一样才叫创新嘛

**Comment by ReactiveCJ** [comment]
> 不一样才叫创新嘛

实际上可以理解，因为mask每一行减去行值，就等于论文中的mask。但是上述代码来自于alibi的原始代码，baichuan把人家注释删了。

**Comment by ReactiveCJ** [comment]
> 不一样才叫创新嘛

抄都能从抄错。
if attention_mask is not None:
            if q_len == 1: # inference with cache
                if len(attention_mask.size()) == 4:
                    attention_mask = attention_mask[:, :, -1:, :]   
                else:
                    # only one head
                    attention_mask = attention_mask[:, -1:, :]   
      
            attn_weights = attn_weights + attention_mask
            attn_weights = torch.max(attn_weights, torch.tensor(torch.finfo(attn_weights.dtype).min))

train状态下，如果有past_key_values, 
attn_weights = attn_weights + attention_mask 这一步是不能加的

**Comment by caiyesd** [comment]
我也注意到这点了，一开始确实有点不能理解。
